[RFC][BPF] Add support for asm gotol_or_nop and nop_or_gotol insns

yonghong-song opened the pull request:
Two new insns are added to BPF instruction set:
  gotol_or_nop
    encoding: gotol encoding with src_reg = 1
  nop_or_gotol
    encoding: gotol encoding with src_reg = 3

Basically src_reg 'bit_0 == 1' means it is gotol_or_nop/nop_or_gotol insn. The src_reg 'bit_1' indicates the insn itself will be a 'goto' 'bit_1 == 0' or a 'nop' 'bit_1 == 1'.

Two insns intend to support kernel static key like transformation where the insn can be a nop or a ja.

The following is an example, where two labels,
static_key_loc_1 and static_key_loc_2, can be used to identify the location of a particular gotol_or_nop/nop_or_gotol location.

It is possible that user space could do
  static_key_enable("static_key_loc_1")
    libbpf can validate that the label "static_key_loc_1" indeed
      corresponds to a gotol_or_nop/nop_or_gotol insn and it
      can translated the 'static_key_enable("static_key_loc_1")'
      to something like bpf syscall command 'static_key_enable prog, insn offset 1'
      and kernel will do proper adjustment.
  the same for static_key_disable, static_key_enabled, etc.

```
$ cat t.c
int bar(void);
int foo1(int arg1)
{
        int a = arg1, b;

        asm volatile goto ("r0 = 0; \
                            static_key_loc_1: \
                            gotol_or_nop %l[label]; \
                            r2 = 2; \
                            r3 = 3; \
                           "::
                            : "r0", "r2", "r3"
                            :label);
        a = bar();
label:
        b = 20 * a;
        return b;
}
int foo2(int arg1)
{
        int a = arg1, b;

        asm volatile goto ("r0 = 0; \
                            static_key_loc_2: \
                            nop_or_gotol %l[label]; \
                            r2 = 2; \
                            r3 = 3; \
                           "::
                            : "r0", "r2", "r3"
                            :label);
        a = bar();
label:
        b = 20 * a;
        return b;
}
$ clang --target=bpf -O2 -g -c t.c
$ llvm-objdump -S t.o
t.o:    file format elf64-bpf

Disassembly of section .text:

0000000000000000 <foo1>:
;       asm volatile goto ("r0 = 0; \
       0:       b7 00 00 00 00 00 00 00 r0 = 0x0

0000000000000008 <static_key_loc_1>:
       1:       06 10 00 00 04 00 00 00 gotol_or_nop +0x4 <LBB0_2>
       2:       b7 02 00 00 02 00 00 00 r2 = 0x2
       3:       b7 03 00 00 03 00 00 00 r3 = 0x3
;       a = bar();
       4:       85 10 00 00 ff ff ff ff call -0x1
       5:       bf 01 00 00 00 00 00 00 r1 = r0

0000000000000030 <LBB0_2>:
;       b = 20 * a;
       6:       27 01 00 00 14 00 00 00 r1 *= 0x14
;       return b;
       7:       bf 10 00 00 00 00 00 00 r0 = r1
       8:       95 00 00 00 00 00 00 00 exit

0000000000000048 <foo2>:
;       asm volatile goto ("r0 = 0; \
       9:       b7 00 00 00 00 00 00 00 r0 = 0x0

0000000000000050 <static_key_loc_2>:
      10:       06 30 00 00 04 00 00 00 nop_or_gotol +0x4 <LBB1_2>
      11:       b7 02 00 00 02 00 00 00 r2 = 0x2
      12:       b7 03 00 00 03 00 00 00 r3 = 0x3
;       a = bar();
      13:       85 10 00 00 ff ff ff ff call -0x1
      14:       bf 01 00 00 00 00 00 00 r1 = r0

0000000000000078 <LBB1_2>:
;       b = 20 * a;
      15:       27 01 00 00 14 00 00 00 r1 *= 0x14
;       return b;
      16:       bf 10 00 00 00 00 00 00 r0 = r1
      17:       95 00 00 00 00 00 00 00 exit
```

inclyc commented on llvm/lib/Target/BPF/BPFInstrInfo.td:
```diff
@@ -604,6 +604,32 @@ class BRANCH_LONG<BPFJumpOp Opc, string OpcodeStr, list<dag> Pattern>
   let BPFClass = BPF_JMP32;
 }
 
+class BRANCH_OR_NOP<BPFJumpOp Opc, string OpcodeStr, list<dag> Pattern>
```
these two insns are just special cases for `JMPL`, how about just extending(inherit) `BRANCH_LONG` and override `Inst{55-52}` fields?

yonghong-song replied:
> these two insns are just special cases for `JMPL`, how about just extending(inherit) `BRANCH_LONG` and override `Inst{55-52}` fields?

Thanks. @inclyc Good suggestion and adding the test! I marked the patch as RFC. Will wait until the design in kernel is settled and then will make proper coding (as you suggested in the above) and add tests.

inclyc reviewed (commented):
I read the thread on https://lore.kernel.org/bpf/20231206141030.1478753-1-aspsk@isovalent.com/T/#mf55ee30dacc6e67aee6712ae3a117059ed06a84b

Looks like there is no test for this, and I just pushed a fixup commit on your branch

inclyc commented on llvm/lib/Target/BPF/AsmParser/BPFAsmParser.cpp:
```diff
@@ -259,6 +261,8 @@ struct BPFOperand : public MCParsedAsmOperand {
         .Case("bswap64", true)
         .Case("goto", true)
         .Case("gotol", true)
+        .Case("gotol_or_nop", true)
```
skeptical about this, is `gotol_or_nop` `ValidIdInMiddle`? I think we did not introduce any asm syntax like:

    if ... gotol_or_nop

yonghong-song replied:
> skeptical about this, is `gotol_or_nop` `ValidIdInMiddle`? I think we did not introduce any asm syntax like:
> 
> ```
> if ... gotol_or_nop
> ```

Thanks! Will remove it later.

Commit:
[BPF] Avoid potential long compilation time without -g

Alastair Robertson reported a huge compilation time increase
without -g for bpf target when comparing to x86 ([1]). In my setup,
with '-O0', for x86, a large basic block compilation takes 0.19s
while bpf target takes 2.46s. The top function which contributes
to the compile time is eliminateFrameIndex().

Such long compilation time without -g is caused by commit
  05de2e481811 ("[bpf] error when BPF stack size exceeds 512 bytes")
The compiler tries to get some debug loc by iterating all insns
in the basic block which will be used when compiler warns
larger-than-512 stack size. Even without -g, such iterating also
happens which cause unnecessary compile time increase.

To fix the issue, let us move the related code when the compiler
is about to warn stack limit violation. This fixed the
compile time regression, and on my system, the compile time
is reduced from 2.46s to 0.35s.

  [1] https://github.com/bpftrace/bpftrace/issues/3257
